### PR TITLE
Add tests to verify existing root component behavior

### PR DESCRIPTION
**Pull request by jvandort**
Add tests to ensure existing root-component-related behavior does not change. The related implementation code will be modified later, so we test edge cases related to the current handling of resolving configurations and what consumable configurations from the same component they have access to.

Also fix two bugs
* Allow users to use `configurations` block in the `buildscript` block. Previously, calls to the `configurations` block would call the same-named method on the parent `Project`. This can be confusing since code which seemingly configures the buildscript's configurations actually configures the Project's configurations
* Ensure that depending on a configuration that is deprecated for consumption emits a warning. previously, a warning would only be emitted if you used a `project(configuration: "X")` dependency if variant aware matching was disabled

### CODE REVIEW

**Comment by jvandort** [comment]
@bot-gradle test this

File: subprojects/dependency-management/src/main/java/org/gradle/internal/component/model/AttributeConfigurationSelector.java [hunk 1]
```diff
@@ -158,7 +159,20 @@ private static List<VariantGraphResolveState> findVariantsProvidingExactlySameCl
...
+        VariantGraphResolveState match = matches.get(0);
+        VariantGraphResolveMetadata matchMetadata = matches.get(0).getMetadata();
+
+        if ((matchMetadata instanceof ConfigurationGraphResolveMetadata) &&
+            ((ConfigurationGraphResolveMetadata) matchMetadata).isDeprecatedForConsumption()
+        ) {
+            DeprecationLogger.deprecateConfiguration(matchMetadata.getName())
+                .forConsumption()
+                .willBecomeAnErrorInGradle9()
+                .withUserManual("dependencies_should_no_longer_be_declared_using_the_compile_and_runtime_configurations")
```
Comment (tresat) [inline subprojects/dependency-management/src/main/java/org/gradle/internal/component/model/AttributeConfigurationSelector.java:172]: ❌   Is this the right user guide section to link to?  This isn't declaration, and how do we know one of these deprecated configurations is involved?  

[The text of this section](https://docs.gradle.org/current/userguide/upgrading_version_5.html#dependencies_should_no_longer_be_declared_using_the_compile_and_runtime_configurations) seems completely unhelpful in this case.

**Review by tresat — CHANGES_REQUESTED** [review]
Changes look good, but I think the documentation link is wrong.

File: subprojects/dependency-management/src/main/java/org/gradle/internal/component/model/AttributeConfigurationSelector.java (hunk 1 above)
Comment (jvandort) [reply subprojects/dependency-management/src/main/java/org/gradle/internal/component/model/AttributeConfigurationSelector.java:172]: This is the same link we already link to for the existing deprecation. Not saying its right but I'm not sure we have a better place to link to right now.

Comment (tresat) [reply subprojects/dependency-management/src/main/java/org/gradle/internal/component/model/AttributeConfigurationSelector.java:172]: @lkasso Do you have any thoughts on if a better section exists in the docs to link this  error to, or if one should be written?

The situation is the second one Justin noted: Depending on and selecting on a configuration that is deprecated for consumption.

Comment (jvandort) [reply subprojects/dependency-management/src/main/java/org/gradle/internal/component/model/AttributeConfigurationSelector.java:172]: The current link is definitely bad. I think this is a sort of unique situation since this deprecation is "ongoing" and isn't necessarily linked to a certain version. There could be many different reasons why a given configuration is deprecated for consumption, so it's difficult to have a link here which genericly works for all configurations.

Comment (lkasso) [reply subprojects/dependency-management/src/main/java/org/gradle/internal/component/model/AttributeConfigurationSelector.java:172]: I've been reviewing the documentation, and we don't have an adequate page for this. The current link is old but it's the best we have for now. I will add this to my to-do list and mention it the #bt-docs slack channel when ready.

File: subprojects/dependency-management/src/integTest/groovy/org/gradle/integtests/resolve/RootComponentResolutionIntegrationTest.groovy [hunk 2]
```diff
@@ -0,0 +1,321 @@
...
+                            artifact file('foo.txt')
+                        }
+                        attributes {
+                            attribute(Category.CATEGORY_ATTRIBUTE, objects.named(Category, "foo"))
+                        }
+                    }
+                }
+
+                dependencies {
+                    conf project(":")
```
[the remark this answers is not in the document]
Comment (jvandort) [reply subprojects/dependency-management/src/integTest/groovy/org/gradle/integtests/resolve/RootComponentResolutionIntegrationTest.groovy:42]: Yeah. I had to do some weird reverse engineering here, but its possible. I wanted this to simply document existing behavior. This isn't necessarily desired behavior, and we likely want to deprecate the behavior displayed in these tests.

File: subprojects/core-api/src/main/java/org/gradle/api/initialization/dsl/ScriptHandler.java [hunk 3]
```diff
@@ -104,6 +105,17 @@ public interface ScriptHandler {
...
+    /**
+     * Configures the configurations for the script. Executes the given closure against the {@link ConfigurationContainer} for
+     * this handler. The {@link ConfigurationContainer} is passed to the closure as the closure's delegate.
+     *
+     * @param configureClosure The closure used to configure the script configurations.
+     *
+     * @since 8.4
+     */
+    @Incubating
+    void configurations(Closure configureClosure);
```
[the remark this answers is not in the document]
Comment (jvandort) [reply subprojects/core-api/src/main/java/org/gradle/api/initialization/dsl/ScriptHandler.java:117]: I tried to do that initially, but there was some weird error (I forget now, I wrote this a while ago). I suspected it has something to do with the fact that this is in the buildscript block and the auto-generation is potentially handled differently. 

I can go back and check what the original error was and see if I can dig deeper. I felt it was not _so_ important, since there are other methods in this class that use the `Closure` style

File: subprojects/dependency-management/src/integTest/groovy/org/gradle/integtests/resolve/RootComponentResolutionIntegrationTest.groovy (hunk 2 above)
[the remark this answers is not in the document]
Comment (big-guy) [reply subprojects/dependency-management/src/integTest/groovy/org/gradle/integtests/resolve/RootComponentResolutionIntegrationTest.groovy:42]: Yeah, I don't expect you to change anything here.  I think there are a few paths on this:

1. Deprecate all of `buildscript {}`
2. Deprecate parts of `buildscript {}` like defining repositories or creating new configurations
3. Restrict `dependencies {}` to a static API (like we have in test suites) where `project(...)` isn't available

[the remark this answers is not in the document]
Comment (jvandort) [reply subprojects/dependency-management/src/integTest/groovy/org/gradle/integtests/resolve/RootComponentResolutionIntegrationTest.groovy:42]: At the very least, I think we should restrict users from being able to create new configurations. The `classpath` configuration is already deprecated for consumption so we'll close that loophole soon too

File: subprojects/core-api/src/main/java/org/gradle/api/initialization/dsl/ScriptHandler.java (hunk 3 above)
[the remark this answers is not in the document]
Comment (jvandort) [reply subprojects/core-api/src/main/java/org/gradle/api/initialization/dsl/ScriptHandler.java:117]: When changed to an Action, the new groovy test fails with:
```
Caused by: org.gradle.internal.metaobject.AbstractDynamicObject$CustomMessageMissingMethodException: Could not find method foo() for arguments [build_29h8tzzwiu6pczslnnnlm1z9l$_run_closure1$_closure2$_closure3@60ccf498] on object of type org.gradle.api.internal.initialization.DefaultScriptHandler.
	at app//org.gradle.internal.metaobject.AbstractDynamicObject$CustomMissingMethodExecutionFailed.<init>(AbstractDynamicObject.java:190)
	at app//org.gradle.internal.metaobject.AbstractDynamicObject.methodMissingException(AbstractDynamicObject.java:184)
	at app//org.gradle.internal.metaobject.ConfigureDelegate.invokeMethod(ConfigureDelegate.java:80)
	at build_29h8tzzwiu6pczslnnnlm1z9l$_run_closure1$_closure2.doCall$original(/Users/jvandort/work/gradle/subprojects/dependency-management/build/tmp/teŝt files/ScriptDepen.Test/cufcu/build.gradle:4)
	at app//org.gradle.api.internal.initialization.DefaultScriptHandler.configurations(DefaultScriptHandler.java:133)
	at app//org.gradle.internal.metaobject.BeanDynamicObject$MetaClassAdapter.invokeMethod(BeanDynamicObject.java:525)
	at app//org.gradle.internal.metaobject.BeanDynamicObject.tryInvokeMethod(BeanDynamicObject.java:216)
	at app//org.gradle.internal.metaobject.ConfigureDelegate.invokeMethod(ConfigureDelegate.java:61)
	at build_29h8tzzwiu6pczslnnnlm1z9l$_run_closure1.doCall$original(/Users/jvandort/work/gradle/subprojects/dependency-management/build/tmp/teŝt files/ScriptDepen.Test/cufcu/build.gradle:3)
	at app//org.gradle.util.internal.ClosureBackedAction.execute(ClosureBackedAction.java:73)
	at app//org.gradle.util.internal.ConfigureUtil.configureTarget(ConfigureUtil.java:155)
	at app//org.gradle.util.internal.ConfigureUtil.configure(ConfigureUtil.java:106)
	at app//org.gradle.api.internal.project.DefaultProject.buildscript(DefaultProject.java:1318)
	at app//org.gradle.api.internal.project.ProjectScript.buildscript(ProjectScript.java:47)
	at build_29h8tzzwiu6pczslnnnlm1z9l.run(/Users/jvandort/work/gradle/subprojects/dependency-management/build/tmp/teŝt files/ScriptDepen.Test/cufcu/build.gradle:2)
	at app//org.gradle.groovy.scripts.internal.DefaultScriptRunnerFactory$ScriptRunnerImpl.run(DefaultScriptRunnerFactory.java:91)
```

It looks like the action method is being called and no Closure method is called. Therefore, it tries to call a method named `foo` instead of going through `ConfigureUtil` to do all the NDOC dynamic groovy magic.

[the remark this answers is not in the document]
Comment (jvandort) [reply subprojects/core-api/src/main/java/org/gradle/api/initialization/dsl/ScriptHandler.java:117]: Instantiating the `DefaultScriptHandler` with an instantiator instead of `new` fixes this.

**Review by lkasso — APPROVED** [review]

**Review by big-guy — APPROVED** [review]

**Comment by jvandort** [comment]
@bot-gradle test and merge
